source/common/event/user_space_file_event_impl.cc

```diff
@@ -0,0 +1,67 @@
+#include "common/event/user_space_file_event_impl.h"
+
+#include <cstdint>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Event {
+
+void DefaultEventListener::onEventEnabled(uint32_t enabled_events) {
+  enabled_events_ = enabled_events;
+  // Clear ephemeral events to align with FileEventImpl::setEnable().
+  ephermal_events_ = 0;
+}
+
+void DefaultEventListener::onEventActivated(uint32_t activated_events) {
+  // Event owner should not activate any event which is disabled.
+  // Also see onEventEnabled which clears ephemeral events.
+  // The overall prevents callback on disabled events.
+  ASSERT((activated_events & ~enabled_events_) == 0);
+  ephermal_events_ |= activated_events;
+}
+
+void UserSpaceFileEventImpl::activate(uint32_t events) {
+  // Only supported event types are set.
+  ASSERT((events & (FileReadyType::Read | FileReadyType::Write | FileReadyType::Closed)) == events);
+  event_listener_.onEventActivated(events);
+  if (!schedulable_.enabled()) {
+    schedulable_.scheduleCallbackNextIteration();
+  }
+}
+
+void UserSpaceFileEventImpl::setEnabled(uint32_t events) {
+  // Only supported event types are set.
+  ASSERT((events & (FileReadyType::Read | FileReadyType::Write | FileReadyType::Closed)) == events);
+  event_listener_.onEventEnabled(events);
```

Comment:
In the future we will need to know if the buffers are readable/writable when setEnabled is called.  What's the best way to provide that information to the UserSpaceFileEvent?

Reply:
Is buffer probing the part of the intended user space file event optimization, or is it ongoing efforts on general file event?

Reply:
I think we should think about minimization of events delivered by UserSpaceFileEventImpl as a correctness issue, not an optimization.

Reply:
Why are you asking a feature that neither libevent nor epoll promises? 
You don't give up linux when the thundering herd problem is more terrible than the situation nowadays, do you?

Reply:
Now read and write should be independent by consulting the ReadWritable (implemented by buffered_io_socket_handle)

Reply:
`In the future we will need to know if the buffers are readable/writable when setEnabled is called.`
My understanding is that this is not a public interface of the event interface. Instead, the goal of the get the enabled events are merely to support `EmulatedEdge`, where simulate Edge trigger events when the OS support only level trigger.

I would argue that user space file event doesn't have to support your mentioned feature. The idea is that this `user space event` is always edge triggered regardless the OS(particularly windows platform) provides an edge poller.